Stratis.Bitcoin/Features/IndexStore/Index.cs

```diff
@@ -0,0 +1,421 @@
+            :base(multiValue, builder, dependencies)
+        {
+            this.repository = repository;
+            this.session = repository.GetSession();
+            this.Name = name;
+            this.Table = repository.IndexTableName(name);
+
+            try
+            {
+                this.Compile();
+            }
+            catch (Exception e)
+            {
+                throw new IndexStoreException("Could not compile index '" + name + "': " + e.Message);
+            }
+        }
+        
+        public static Index Parse(IndexRepository repository, string json, string table = null)
+        {
+            var index = new Index(repository);
+
+            JsonConvert.PopulateObject(json, index);
+            if (table != null)
+                index.Table = table;
+
+            try
+            {
+                index.Compile();
+            }
+            catch (Exception e)
+            {
+                throw new IndexStoreException("Could not compile index '" + index.Name + "': " + e.Message);
+            }
+
+            return index;
+        }
+
+        public override string ToString()
+        {
+            return JsonConvert.SerializeObject(this);
+        }
+
+        public delegate void recordIndexed(object sender, byte[] key, byte[] value);
+
+        public event recordIndexed recordAdded;
+        public event recordIndexed recordRemoved;
+
+        public IEnumerable<byte[]> LookupMultiple(List<byte[]> keyList)
+        {
+            if (this.Many)
+                throw new IndexStoreException("Index '" + this.Name + "' can't be multi-value");
+
+            foreach (var key in keyList)
+            {
+                if (key != null)
+                {
+                    var trxId = this.session.Transaction.Select<byte[], byte[]>(this.Table, key);
+
+                    if (trxId.Exists)
+                    {
+                        this.PerformanceCounter.AddRepositoryHitCount(1);
+                        yield return trxId.Value;
+                        continue;
+                    }
+                }
+
+                this.PerformanceCounter.AddRepositoryMissCount(1);
+                yield return null;
+            }
+        }
+
+        public IEnumerable<byte[]> EnumerateValues(byte[] key)
+        {
+            if (!this.Many)
+                throw new IndexStoreException("Index '" + this.Name + "' should be multi-value");
+
+            var addrRow = this.session.Transaction.Select<byte[], byte[]>(this.Table, key);
+            if (addrRow.Exists)
+            {
+                // Get the current blob size
+                var sizeBytes = addrRow.GetValuePart(0, sizeof(uint));
+                if (BitConverter.IsLittleEndian) sizeBytes = sizeBytes.Reverse();
+                uint size = BitConverter.ToUInt32(sizeBytes, 0);
+
+                for (uint offset = 0; offset < size;)
```

Comment:
"InsertValue", "DeleteValues" and "EnumerateValue" all relate to multi-value (many) indexes. As you  have noticed a multi-value index has multiple values for the same key stored in a sequence of segments consisting of the value's length followed by the value. The whole sequence also has a size associated with it which can be less than the size recorded by DBreeze for the "value". We grow that size in exponents of 2 to reduce the number of times we have to do that.